src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -25,13 +26,33 @@ internal class DefaultRazorLanguageServerCustomMessageTarget : RazorLanguageServ
         private readonly JoinableTaskFactory _joinableTaskFactory;
         private readonly LSPRequestInvoker _requestInvoker;
         private readonly RazorUIContextManager _uIContextManager;
+        private readonly IDisposable _razorReadyListener;
+
+        private const string RazorReadyFeature = "Razor-Initialization";
 
         [ImportingConstructor]
         public DefaultRazorLanguageServerCustomMessageTarget(
             LSPDocumentManager documentManager,
             JoinableTaskContext joinableTaskContext,
             LSPRequestInvoker requestInvoker,
-            RazorUIContextManager uIContextManager)
+            RazorUIContextManager uIContextManager,
+            IRazorAsynchronousOperationListenerProviderAccessor asyncOpListenerProvider) :
+                this(
+                    documentManager,
+                    joinableTaskContext,
+                    requestInvoker,
+                    uIContextManager,
+                    asyncOpListenerProvider.GetListener(RazorReadyFeature).BeginAsyncOperation(RazorReadyFeature))
```

Comment:
I'm having a hard time piecing all of the pieces of the puzzle together. So Roslyn provides an [AsynchronousOperationListenerProvider](https://github.com/dotnet/roslyn/blob/c9a68b6a2fed69b1a48505b401dd1bb3d09dc45a/src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListenerProvider.cs) which allows people ([us](https://github.com/dotnet/roslyn/pull/50970)) to register asynchronous events that can be be waited on right? Does the platform integrate with this asynchronous provider on the host box via the "background tasks running" dialog in the bottom left corner of VS which indirectly tells the DDRITs platform to wait for all of those tasks to be "done"?

Reply:
CC @sharwell to correct me if I'm wrong.

My understanding of the flow here is that AsynchronousOperationListenerProvider creates a BrokeredService* which will create Async objects that can be awaited on. Those async objects are resolved when the `IDisposable` returned by `BeginAsyncOperation` is disposed. To my knowledge it  doesn't involve the "Background tasks running" dialog.

Then on the DDRIT side we call [IApexAsynchronousOperationListenerProviderAccessor](https://github.com/dotnet/roslyn/blob/c9a68b6a2fed69b1a48505b401dd1bb3d09dc45a/src/Tools/ExternalAccess/Apex/IApexAsynchronousOperationListenerProviderAccessor.cs).WaitAll, which Brokers things to give us an awaitable object that resolves when the Operation is done.

* AFAIK "BrokeredService" is a VS concept which will either perform the operation on the local (be it client or server) or make a network call out to the remote (client or server) to have them perform the operation.